### PR TITLE
Add max_depth parameter to ModelSummary

### PR DESCRIPTION
**Pull request by ManuelPalermo**
## What does this PR do?

Adds feature discussed in https://github.com/PyTorchLightning/pytorch-lightning/issues/8047#issue-925641509

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -1610,11 +1610,13 @@ def tbptt_split_batch(self, batch, split_size):
 
         return splits
 
-    def summarize(self, mode: Optional[str] = ModelSummary.MODE_DEFAULT) -> Optional[ModelSummary]:
+    def summarize(self,
+                  mode: Optional[str] = ModelSummary.MODE_DEFAULT,
```
Comment (kaushikb11) [inline pytorch_lightning/core/lightning.py:1614]:
```suggestion
                  mode: str = ModelSummary.MODE_DEFAULT,
```

File: pytorch_lightning/core/memory.py [hunk 2]
```diff
@@ -137,6 +137,8 @@ class ModelSummary(object):
              - `top` (default): only the top-level modules will be recorded (the children of the root module)
              - `full`: summarizes all layers and their submodules in the root module
 
+        max_depth: Maximum depth of modules to show when mode="full"
```
Comment (awaelchli) [inline pytorch_lightning/core/memory.py:140]: If we really want `max_depth`, we would have to drop `mode` I'm pretty sure because two arguments that have overlapping functionality is not good to maintain, i.e., `mode="top"` is the same as `max_depth=1` and `mode="full"` is the same as `max_depth = -1`.  

I propose the following: Drop `mode` on  `ModelSummary` completely. Map `Trainer.weights_summary=full|top|None` and `Lightning.summarize` to the appropriate value of `max_depth`.

Comment (awaelchli) [reply pytorch_lightning/core/memory.py:140]: this would also avoid your error handling below

File: pytorch_lightning/core/lightning.py [hunk 3]
```diff
@@ -1610,11 +1610,13 @@ def tbptt_split_batch(self, batch, split_size):
 
         return splits
 
-    def summarize(self, mode: Optional[str] = ModelSummary.MODE_DEFAULT) -> Optional[ModelSummary]:
+    def summarize(self,
+                  mode: str = ModelSummary.MODE_DEFAULT,
+                  max_depth: Optional[int] = None) -> Optional[ModelSummary]:
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:1615]:
```suggestion
                  max_depth: Optional[int] = None,) -> Optional[ModelSummary]:
```

File: tests/core/test_memory.py [hunk 4]
```diff
@@ -336,3 +359,16 @@ def test_lazy_model_summary():
             # https://github.com/pytorch/pytorch/issues/58350
             assert summary.total_parameters == 7
             assert summary.trainable_parameters == 7
+
+
+def test_max_depth_0_equals_mode_top():
+    assert str(DeepNestedModel().summarize(mode="top")) \
+           == str(DeepNestedModel().summarize(mode="full", max_depth=0))
```
Comment (Borda) [inline tests/core/test_memory.py:366]:
```suggestion
    summary_top = DeepNestedModel().summarize(mode="top")
    summary_full_0 = DeepNestedModel().summarize(mode="full", max_depth=0)
    assert str(summary_top) == str(summary_full_0)
```

File: pytorch_lightning/core/memory.py (hunk 2 above)
Comment (ManuelPalermo) [reply pytorch_lightning/core/memory.py:140]: Yes, I agree, although I did not want to make a breaking change. Your proposal seems to resolve things nicely if I understand correctly. I just have some questions regarding the implementation:

* Should I leave the `ModelSummary.MODES`? this is used in checks in `Lightning.summarize()`, `Trainer.weights_summary` and in the pytests i.e., `if mode in ModelSummary.MODES:`. Or do I just replace the checks with: `if mode in ("top", "full"):`)? or move the variables elsewhere?
* To make full use of the max_depth functionality, the user would be required to use the `ModelSummary` class directly, leaving the old interface with `mode` in the `Trainer` and `Lightning.summarize` right?

Comment (kaushikb11) [reply pytorch_lightning/core/memory.py:140]: @awaelchli How about having an argument that could support both?

Comment (awaelchli) [reply pytorch_lightning/core/memory.py:140]: Forgot to say thanks for sending a PR and your ideas! Greatly appreciated. 

It would be a change but not a serious one. We would of course not immediately remove it but deprecate the argument. And since we would leave the trainer flag intact and only touch ModelSummary directly, not many users are affected :) 

I suggest we stick to the contents of this PR, with the addition of adding a deprecation message. Then, in a following PR re-route the old mode argument to max_depth (and support -1) and update tests. Until deprecation is done, `ModelSummary.MODES` can stay or be hardcoded and eventually be removed.

> To make full use of the max_depth functionality, the user would be required to use the ModelSummary class directly, leaving the old interface with mode in the Trainer and Lightning.summarize right?

That's actually how many features in Lightning are controlled. A simple on/off switch in the Trainer for convenience and for customization we would interact with the feature instances directly, e.g., loggers callbacks profilers etc. So I think that pattern for model summary fits well. `max_depth` could still remain an argument to summarize() as you have it now.

Of course, other reviewers may also have some input.

Comment (ManuelPalermo) [reply pytorch_lightning/core/memory.py]: I've already added the mapping from `mode` (in Trainer / PL.summarize) to max_depth to the `ModelSummary` and updated the tests accordingly.

The `PL.summarize()` can work with `mode` or `max_depth` parameters, but will raise a warning if only `mode` is passed.

Let me know if you agree.

**Review by tchaton — APPROVED** [review]
LGTM ! Nice contribution.

File: pytorch_lightning/core/memory.py [hunk 5]
```diff
@@ -132,10 +132,7 @@ class ModelSummary(object):
 
     Args:
         model: The model to summarize (also referred to as the root module)
-        mode: Can be one of
-
-             - `top` (default): only the top-level modules will be recorded (the children of the root module)
-             - `full`: summarizes all layers and their submodules in the root module
+        max_depth: Maximum depth of modules to show, or -1 to show all modules.
```
Comment (carmocca) [inline pytorch_lightning/core/memory.py:135]: So a 0 based value, right? Can you mention it?

File: pytorch_lightning/core/memory.py [hunk 6]
```diff
@@ -182,14 +179,11 @@ class ModelSummary(object):
...
     """
 
-    MODE_TOP = "top"
-    MODE_FULL = "full"
-    MODE_DEFAULT = MODE_TOP
-    MODES = [MODE_FULL, MODE_TOP]
+    MODES = dict(top=0, full=-1)
 
-    def __init__(self, model, mode: str = MODE_DEFAULT):
+    def __init__(self, model, max_depth: int = -1):
```
Comment (carmocca) [inline pytorch_lightning/core/memory.py:184]: I believe this class can be used by itself

cc: @awaelchli 

So maybe the previous init argument should be deprecated too

File: pytorch_lightning/core/lightning.py [hunk 7]
```diff
@@ -1610,14 +1610,24 @@ def tbptt_split_batch(self, batch, split_size):
...
 
-        if mode in ModelSummary.MODES:
-            model_summary = ModelSummary(self, mode=mode)
-            log.info("\n" + str(model_summary))
-        elif mode is not None:
-            raise MisconfigurationException(f"`mode` can be None, {', '.join(ModelSummary.MODES)}, got {mode}")
+        if max_depth is None:
+            rank_zero_deprecation(
+                "The use of model.summarize(mode) argument will be "
+                "deprecated in a future release, in favour of 'max_depth'."
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:1619]: This should mention when it was deprecated and when it will be removed

See for example another deprecation PR: https://github.com/PyTorchLightning/pytorch-lightning/pull/7168/files

File: pytorch_lightning/trainer/trainer.py [hunk 8]
```diff
@@ -941,7 +941,8 @@ def _pre_training_routine(self):
 
         # print model summary
         if self.is_global_zero and self.weights_summary is not None and not self.testing:
-            ref_model.summarize(mode=self.weights_summary)
+            max_depth = ModelSummary.MODES[self.weights_summary]
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:944]: Should the Trainer.weights_summary typing signature be updated to allow int?

And should we deprecate passing a str for it?

File: pytorch_lightning/core/lightning.py (hunk 7 above)
Comment (ManuelPalermo) [reply pytorch_lightning/core/lightning.py:1619]: Which versions do you want to target for deprecation / removal?  v1.4 and  v1.5?

"The use of `model.summarize(mode)` argument is deprecated in v1.4 and will be removed in v1.5, use `max_depth` argument instead"

File: pytorch_lightning/trainer/trainer.py (hunk 8 above)
Comment (ManuelPalermo) [reply pytorch_lightning/trainer/trainer.py:944]: I did not make the changes to `Trainer.weights_summary` following https://github.com/PyTorchLightning/pytorch-lightning/pull/8062#discussion_r655532083 . But it might be less confusing to change everything to an int based argument at once, making it consistent.

Let me know what you prefer.

File: pytorch_lightning/core/memory.py (hunk 6 above)
Comment (ManuelPalermo) [reply pytorch_lightning/core/memory.py:184]: yes, the `ModelSummary` class can be used by itself and I've removed mentions to the `mode` argument, except for `    MODES = dict(top=0, full=-1)` which is still used outside to route the old mode argument and to maintain existing checks.

File: pytorch_lightning/core/memory.py (hunk 5 above)
Comment (ManuelPalermo) [reply pytorch_lightning/core/memory.py:135]: I'll update it on the next commit along with other suggested changes:
```python
"""
    (...)
    max_depth: Maximum depth of modules to show, starting on 0 (for modules connected root),
                or -1 to show all modules.
    (...)
"""
```

File: pytorch_lightning/core/memory.py (hunk 6 above)
Comment (ananthsub) [reply pytorch_lightning/core/memory.py:184]: @ManuelPalermo this would be a breaking change, so we can't outright change the parameter like this. it should go through a similar deprecation process

**Review by ananthsub — CHANGES_REQUESTED** [review]

File: pytorch_lightning/core/lightning.py [hunk 9]
```diff
@@ -1610,14 +1610,24 @@ def tbptt_split_batch(self, batch, split_size):
 
         return splits
 
-    def summarize(self, mode: Optional[str] = ModelSummary.MODE_DEFAULT) -> Optional[ModelSummary]:
+    def summarize(self, mode: Optional[str] = "top", max_depth: Optional[int] = None) -> Optional[ModelSummary]:
         model_summary = None
```
Comment (ananthsub) [inline pytorch_lightning/core/lightning.py:1614]: unrelated to this PR, but I feel `summarize` shouldn't be a method on the core lightning module API. it seems like a utility function that accepts the LightningModule as input along with the mode/depth would be preferable instead

File: pytorch_lightning/core/memory.py (hunk 6 above)
Comment (ManuelPalermo) [reply pytorch_lightning/core/memory.py:184]: I was following the suggestion from https://github.com/PyTorchLightning/pytorch-lightning/pull/8062#discussion_r655456826

Comment (awaelchli) [reply pytorch_lightning/core/memory.py:184]: yes, by dropping it I meant remove it but of course deprecate first, then remove ^^ 
simply remapping mode to max_depth during deprecation phase :)

Comment (ManuelPalermo) [reply pytorch_lightning/core/memory.py:184]: ah, ok, sorry for the misunderstanding! I'll update the PR

**Comment by awaelchli** [comment]
@ManuelPalermo I added a couple of commits related to the deprecation stuff. 

One other thing is that currently max_depth = 0 is showing the children for the root module. But it should be
max_depth = 0: No summary
max_depth = 1: Children (mode="top")
etc.

File: pytorch_lightning/core/lightning.py (hunk 9 above)
Comment (carmocca) [reply pytorch_lightning/core/lightning.py:1646]: Related to your comment @ananthsub

https://github.com/PyTorchLightning/pytorch-lightning/issues/4541

File: pytorch_lightning/core/lightning.py [hunk 10]
```diff
@@ -1642,14 +1642,23 @@ def tbptt_split_batch(self, batch, split_size):
...
-            model_summary = ModelSummary(self, mode=mode)
-            log.info("\n" + str(model_summary))
-        elif mode is not None:
-            raise MisconfigurationException(f"`mode` can be None, {', '.join(ModelSummary.MODES)}, got {mode}")
+        # temporary mapping from mode to max_depth
+        if max_depth is None:
+            if mode in ModelSummary.MODES:
+                rank_zero_deprecation(
+                    "Argument `mode` in `LightningModule.summarize` is deprecated in v1.4"
+                    " and will be removed in v1.5. Use `max_depth` instead."
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:1653]: This should recommend the specific `max_depth` value for the `mode` provided

File: pytorch_lightning/core/memory.py [hunk 11]
```diff
@@ -249,6 +272,12 @@ def summarize(self) -> Dict[str, LayerSummary]:
             self._forward_example_input()
         for layer in summary.values():
             layer.detach_hook()
+
+        if self._max_depth >= 1:
+            # remove summary entries with depth > max_depth
+            for k in [k for k in summary.keys() if k.count(".") >= self._max_depth]:
```
Comment (carmocca) [inline pytorch_lightning/core/memory.py:278]:
```suggestion
            for k in [k for k in summary if k.count(".") >= self._max_depth]:
```

File: pytorch_lightning/core/memory.py [hunk 12]
```diff
@@ -182,14 +189,25 @@ class ModelSummary(object):
...
+            if mode in ModelSummary.MODES:
+                from pytorch_lightning.utilities import rank_zero_deprecation
+                rank_zero_deprecation(
+                    "Argument `mode` in `ModelSummary` is deprecated in v1.4"
+                    " and will be removed in v1.5. Use `max_depth` instead."
+                )
+                max_depth = ModelSummary.MODES[mode]
+            else:
+                from pytorch_lightning.utilities.exceptions import MisconfigurationException
+                raise MisconfigurationException(f"`mode` can be {', '.join(ModelSummary.MODES)}, got {mode}.")
```
Comment (carmocca) [inline pytorch_lightning/core/memory.py:208]: To mimic the previous behaviour, this should set `max_depth = 1` when the user does not pass any argument

**Review by ananthsub — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Review by awaelchli — CHANGES_REQUESTED** [review]
this doesn't work yet. 

`Trainer(weights_summary="full"|"top")` does not display anything on this branch. 

need to investigate.

**Review by awaelchli — APPROVED** [review]
